tsl/src/nodes/vector_agg/grouping_policy_hash.c

```diff
@@ -0,0 +1,735 @@
+ */
+
+/*
+ * This grouping policy groups the rows using a hash table. Currently it only
+ * supports a single fixed-size by-value compressed column that fits into a Datum.
+ */
+
+#include <postgres.h>
+
+#include <executor/tuptable.h>
+#include <nodes/pg_list.h>
+
+#include "grouping_policy.h"
+
+#include "nodes/decompress_chunk/compressed_batch.h"
+#include "nodes/vector_agg/exec.h"
+
+#ifdef USE_FLOAT8_BYVAL
+#define DEBUG_LOG(MSG, ...) elog(DEBUG3, MSG, __VA_ARGS__)
+#else
+/*
+ * On 32-bit platforms we'd have to use the cross-platform int width printf
+ * specifiers which are really unreadable.
+ */
+#define DEBUG_LOG(...)
+#endif
+
+/*
+ * We can use crc32 as a hash function, it has bad properties but takes only one
+ * cycle, which is why it is sometimes used in the existing hash table
+ * implementations.
+ */
+#ifdef USE_SSE42_CRC32C
+#include <nmmintrin.h>
+static pg_attribute_always_inline uint64
+hash64(uint64 x)
+{
+	return _mm_crc32_u64(~0ULL, x);
+}
+
+#else
+/*
+ * When we don't have the crc32 instruction, use the SplitMix64 finalizer.
+ */
+static pg_attribute_always_inline uint64
+hash64(uint64 x)
+{
+	x ^= x >> 30;
+	x *= 0xbf58476d1ce4e5b9U;
+	x ^= x >> 27;
+	x *= 0x94d049bb133111ebU;
+	x ^= x >> 31;
+	return x;
+}
+#endif
+
+/*
+ * For the hash table, use the generic Datum key that is mapped to the aggregate
+ * state index.
+ */
+typedef struct
+{
+	Datum key;
+	uint32 status;
+	uint32 agg_state_index;
+} HashEntry;
+
+#define SH_PREFIX h
+#define SH_ELEMENT_TYPE HashEntry
+#define SH_KEY_TYPE Datum
+#define SH_KEY key
+#define SH_HASH_KEY(tb, key) hash64(key)
+#define SH_EQUAL(tb, a, b) a == b
+#define SH_SCOPE static inline
+#define SH_DECLARE
+#define SH_DEFINE
+#include <lib/simplehash.h>
+
+/*
+ * Hash grouping policy.
+ */
+typedef struct
+{
+	/*
+	 * We're using data inheritance from the GroupingPolicy.
+	 */
+	GroupingPolicy funcs;
+
+	List *agg_defs;
+	List *output_grouping_columns;
+
+	/*
+	 * The hash table we use for grouping.
+	 */
+	struct h_hash *table;
+
+	/*
+	 * We have to track whether we are in the mode of returning the partial
+	 * aggregation results, and also use a hash table iterator to track our
+	 * progress between emit() calls.
+	 */
+	bool returning_results;
+	struct h_iterator iter;
+
+	/*
+	 * In single-column grouping, we store the null key outside of the hash
+	 * table, and it has a reserved aggregate state index 1. We also reset this
+	 * flag after we output the null key during iteration.
+	 */
+	bool have_null_key;
+
+	/*
+	 * A memory context for aggregate functions to allocate additional data,
+	 * i.e. if they store strings or float8 datum on 32-bit systems. Valid until
+	 * the grouping policy is reset.
+	 */
+	MemoryContext agg_extra_mctx;
+
+	/*
+	 * Temporary storage of aggregate state offsets for a given batch. We keep
+	 * it in the policy because it is potentially too big to keep on stack, and
+	 * we don't want to reallocate it each batch.
+	 */
```

Comment:
Took me a while to figure out what these offsets are for. Could we do a better description here?

For example (if I got things right):

The agg_state_offsets array  contains an offset for each "row" in an ArrowArray that points to the aggregate states for that row. The agg state is points to is decided by the value (key) in the grouping column. Thus, the number of agg states is equal to the number of unique values in the grouping column's ArrowArray.

Reply:
Yeah I'll try to describe it better. They are basically unique indexes of the grouping keys, and the array is indexed by the row index of the input batch.